cmd/tracee-ebpf/main.go

```diff
@@ -985,13 +983,61 @@ func prepareEventsToTrace(eventFilter *tracee.StringFilter, setFilter *tracee.St
 	return res, nil
 }
 
-func checkRequiredCapabilities(caps capability.Capabilities) error {
-	if !caps.Get(capability.EFFECTIVE, capability.CAP_SYS_ADMIN) {
-		return fmt.Errorf("insufficient privileges to run: missing CAP_SYS_ADMIN")
+const bpfCapabilitiesMinKernelVersion = "5.8"
+
+// ensureCapabilities makes sure the program has just the required capabilities to run
+func ensureCapabilities(OSInfo *helpers.OSInfo) error {
+	selfCap, err := getSelfCapabilities()
+	if err != nil {
+		return err
 	}
 
-	if !caps.Get(capability.EFFECTIVE, capability.CAP_IPC_LOCK) {
-		return fmt.Errorf("insufficient privileges to run: missing CAP_IPC_LOCK")
+	// Build the set of capabilities required to run
+	rCaps := []capability.Cap{
+		capability.CAP_IPC_LOCK,
+		capability.CAP_SYS_PTRACE,
+		capability.CAP_SYS_RESOURCE,
+		capability.CAP_NET_ADMIN,
+	}
+	if OSInfo.CompareOSBaseKernelRelease(bpfCapabilitiesMinKernelVersion) <= 0 {
+		rCaps = append(rCaps, []capability.Cap{
+			capability.CAP_BPF,
+			capability.CAP_PERFMON,
+		}...)
+	} else {
+		rCaps = append(rCaps, []capability.Cap{
+			capability.CAP_SYS_ADMIN,
+		}...)
+	}
+
+	if err = checkRequiredCapabilities(selfCap, rCaps); err != nil {
+		return err
+	}
+	if err = dropUnrequiredCapabilities(selfCap, rCaps); err != nil {
+		return err
+	}
+	return nil
+}
+
+// dropUnrequiredCapabilities requires that all capabilities are already set, or SYS_SETPCAP capability is set
+// The function also tries to drop the capabilities bounding set, but it won't work if SYS_SETPCAP is not available.
```

Comment:
I believe you meant CAP_SETPCAP here, not SYS_, right ? As I'm reading, CAP_SETPCAP might not be available IF file capabilities is, correct ? Should we inform the user of one or another or selfCaps.Apply will handle the error ?

Reply:
The use I do with CAP_SETPCAP is just a little extra I added to take advantage of extra capabilities to make the process even more secured. This capability is not a must-have, rather a nice-to-have. I can print a log that the capability wasn't present, but I don't think that it is really important to have.

Reply:
I didn't add message to the user if the capability is not preset because I fear it might be a spam.
Tell me if you want me to add it either way.